apps/astarte_realm_management/lib/astarte_realm_management/engine.ex

```diff
@@ -273,43 +273,64 @@ defmodule Astarte.RealmManagement.Engine do
     end
   end
 
-  # TODO: Mappings documentation changes are discarded
-  defp extract_new_mappings(db_client, %{mappings: upd_mappings} = interface_doc) do
+  defp extract_updated_mappings(db_client, %{mappings: upd_mappings} = interface_doc) do
     descriptor = InterfaceDescriptor.from_interface(interface_doc)
 
-    with {:ok, mappings} <- Mappings.fetch_interface_mappings(db_client, descriptor.interface_id) do
+    with {:ok, mappings_map} <-
+           Mappings.fetch_interface_mappings_map(db_client, descriptor.interface_id,
+             include_docs: true
+           ) do
       upd_mappings_map =
         Enum.into(upd_mappings, %{}, fn mapping ->
           {mapping.endpoint_id, mapping}
         end)
 
-      maybe_new_mappings =
-        Enum.reduce_while(mappings, upd_mappings_map, fn mapping, acc ->
-          case drop_mapping_doc(Map.get(upd_mappings_map, mapping.endpoint_id)) do
-            nil ->
-              {:halt, {:error, :missing_endpoints}}
-
-            ^mapping ->
-              {:cont, Map.delete(acc, mapping.endpoint_id)}
+      existing_endpoints = Map.keys(mappings_map)
+      {existing_mappings, new_mappings} = Map.split(upd_mappings_map, existing_endpoints)
 
-            _ ->
-              {:halt, {:error, :incompatible_endpoint_change}}
-          end
-        end)
+      with {:ok, updated_mappings} <- extract_changed_mappings(mappings_map, existing_mappings) do
+        {:ok, Map.merge(new_mappings, updated_mappings)}
+      end
+    end
+  end
 
-      if is_map(maybe_new_mappings) do
-        {:ok, maybe_new_mappings}
+  defp extract_changed_mappings(old_mappings, existing_mappings) do
+    Enum.reduce_while(old_mappings, {:ok, %{}}, fn {mapping_id, old_mapping}, {:ok, acc} ->
+      with {:ok, updated_mapping} <- Map.fetch(existing_mappings, mapping_id),
+           {:allowed, true} <- {:allowed, allowed_mapping_update?(updated_mapping, old_mapping)},
+           {:updated, true} <- {:updated, is_mapping_updated?(updated_mapping, old_mapping)} do
+        {:cont, {:ok, Map.put(acc, mapping_id, updated_mapping)}}
       else
-        maybe_new_mappings
+        :error ->
+          {:halt, {:error, :missing_endpoints}}
+
+        {:allowed, false} ->
+          {:halt, {:error, :incompatible_endpoint_change}}
+
+        {:updated, false} ->
+          {:cont, {:ok, acc}}
       end
-    end
+    end)
+  end
+
+  defp allowed_mapping_update?(upd_mapping, mapping) do
```

Comment:
Since you're passing `old_mappings, existing_mappings` to `extract_changed_mappings` I'd be consistent and alway pass the old one first